Reduce docker image size (and build time)

lcobucci opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | yes

### Description

This improves the build process of the docker image.

When the base images are already downloaded, it takes about 15s to build the image using BuildKit - previously it was taking 1m19s.
The image size has also been drastically reduced, it's 122Mb - previously was 390Mb.

lcobucci committed:
Use official PHP image based on Alpine Linux

This swaps the base image so that we can have:
* faster builds
* smaller images
* more secure and up-to-date dependencies

Signed-off-by: Luís Cobucci <lcobucci@gmail.com>

lcobucci committed:
Use COPY over ADD

It's recommended to use `COPY` instead of `ADD` because it's much
simpler and avoids unexpected behaviour - `ADD` can also do things like
uncompressing files and remote requests.

More info: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/#add-or-copy

Signed-off-by: Luís Cobucci <lcobucci@gmail.com>

lcobucci committed:
Avoid trying to re-install dependencies

We only need to dump the autoloader at this stage, so it's unnecessary
to do anything beyond that.

Signed-off-by: Luís Cobucci <lcobucci@gmail.com>

lcobucci commented:
We can compare the speed improvements on the build too: [a previous run took 1m31s](https://github.com/laminas/automatic-releases/pull/102/checks?check_run_id=1763804265) while [this one took 41s](https://github.com/laminas/automatic-releases/pull/115/checks?check_run_id=1974463684).

lcobucci commented:
I just noticed that we don't have the branch `1.10.x` and since this isn't a bug fix it shouldn't be merged to `1.9.x`. I'll convert it into a draft for now - to avoid accidental merges.

weierophinney commented on Dockerfile:
```diff
@@ -15,33 +15,21 @@ LABEL "maintainer"="https://github.com/laminas/technical-steering-committee/"
+COPY src /app/src/
 
-RUN composer install -a --no-dev
+RUN composer dumpautoload -a --no-dev
```
Why `dumpautoload` and not `install`?

lcobucci replied:
I've tried to cover that on the [commit message](https://github.com/laminas/automatic-releases/pull/115/commits/3aba15503c21b44575cdf92f42f77ecb98ed6795). But the dependencies are already installed at this point, we just need to create the autoloader.

Ocramius reviewed (approved):
LGTM! I just need to do the branching ballet to get this released then :-)

Slamdunk commented on Dockerfile:
```diff
@@ -15,33 +15,21 @@ LABEL "maintainer"="https://github.com/laminas/technical-steering-committee/"
+COPY src /app/src/
 
-RUN composer install -a --no-dev
+RUN composer dumpautoload -a --no-dev
```
```suggestion
RUN composer dump-autoload -a --no-dev
```

See https://github.com/composer/composer/blob/e3af4ee606023079d8af222da82a5b6f36ca879f/src/Composer/Command/DumpAutoloadCommand.php#L29-L30

lcobucci committed:
Build image using BuildKit

BuildKit is an optimised builder for docker.

Although it isn't enabled by default, it helps to speed up the process
even more and it's quite stable at this point.

More info: https://docs.docker.com/develop/develop-images/build_enhancements/

Signed-off-by: Luís Cobucci <lcobucci@gmail.com>

Ocramius commented on .github/workflows/docker-build.yml:
```diff
@@ -18,4 +18,4 @@ jobs:
 
       - name: "Docker Build"
-        run: "docker build ."
+        run: "DOCKER_BUILDKIT=1 docker build ."
```
BUILDKIT FTW :muscle:

Ocramius commented:
Branching fixed: :ship:

lcobucci commented:
Thanks @Ocramius

weierophinney commented:
I CANNOT WAIT for my next release to see the speed improvements!

Ocramius commented:
@weierophinney FWIW, 43s runtime: https://github.com/laminas/automatic-releases/runs/1978710568?check_suite_focus=true
